File: examples/google-wallet-instant-app/instantapp/src/main/java/br/com/suaempresa/identificacao/NfcBeam.kt
package br.com.suaempresa.identificacao

import android.nfc.NdefMessage
import android.nfc.NdefRecord

/**
 * Empacota o transport_field num NDEF message para envio por NFC ao
 * terminal do pagador.
 *
 * Modelo de entrega no Android:
 *   - HCE (Host Card Emulation): o device do recebedor emula um tag
 *     NFC; o terminal do pagador (reader) lê ao aproximar. Exige um
 *     `HostApduService` declarado no manifest.
 *   - NDEF push direto: descontinuado (Android Beam removido no
 *     Android 10). HCE é o caminho atual.
 *
 * ⚠️ CAVEAT Instant App + NFC: `android.permission.NFC` é concedida
 * sem prompt, mas registrar um HostApduService a partir de um Instant
 * App tem limites de ciclo de vida. O caminho robusto continua sendo
 * o QR óptico (QrCode.kt); o NFC é conveniência "tap" complementar.
 *
 * Este helper só constrói o NDEF message; a integração com o
 * HostApduService (resposta a APDUs SELECT/READ) fica a cargo do
 * integrador conforme o protocolo do terminal do pagador.
 */
object NfcBeam {

    /** MIME type customizado que o terminal do pagador filtra. */
    private const val MIME_TYPE = "application/vnd.mcpix.transport"

    /**
     * Constrói o NDEF message carregando o transport_field como um
     * record MIME. O leitor (pagador) reconhece o MIME type e o
     * prefixo PIXOFFv1 do payload, roteando para o fluxo mcpix.
     */
    fun buildNdef(transportField: String): NdefMessage {
        val record = NdefRecord.createMime(
            MIME_TYPE,
            transportField.toByteArray(Charsets.US_ASCII),
        )
        return NdefMessage(arrayOf(record))
    }
}

File: examples/apple-wallet-appclip/AppClip/NFCBeam.swift
// Envio do transport_field por NFC ao encostar no terminal do pagador.
//
// Modelo: o App Clip do recebedor escreve o `transport_field` num NDEF
// message; o terminal do pagador (que está em modo leitura NFC) o lê
// ao aproximar. Usa CoreNFC — framework nativo.
//
// ⚠️ CAVEAT App Clip + NFC (ver README §"NFC"):
//   - CoreNFC exige o entitlement
//     `com.apple.developer.nfc.readersession.formats`.
//   - App Clips suportam um SUBSET de entitlements. NFC tag *reading*
//     é suportado; peer-to-peer NDEF *writing* tem suporte que varia
//     por device/iOS. Confirme antes de depender só do NFC.
//   - O QR óptico (BarcodeView) é o caminho primário robusto; o NFC
//     é conveniência "tap" complementar.
//
// Este arquivo modela o lado RECEBEDOR escrevendo num NDEF tag/peer.
// Em iOS, escrita NDEF é feita sobre um `NFCNDEFTag` descoberto numa
// `NFCNDEFReaderSession` — o "tag" pode ser outro device em modo
// emulação ou um NFC sticker reprogramável no terminal do pagador.

import CoreNFC

final class NFCBeam: NSObject, NFCNDEFReaderSessionDelegate {
    private var session: NFCNDEFReaderSession?
    private var payload: String = ""
    private var onResult: ((Result<Void, Error>) -> Void)?

    /// Inicia uma sessão NFC e escreve `transportField` no primeiro
    /// tag NDEF aproximado.
    func beam(transportField: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard NFCNDEFReaderSession.readingAvailable else {
            completion(.failure(NFCError.unavailable))
            return
        }
        self.payload = transportField
        self.onResult = completion
        let session = NFCNDEFReaderSession(
            delegate: self,
            queue: nil,
            invalidateAfterFirstRead: true
        )
        session.alertMessage = "Aproxime do terminal de pagamento"
        session.begin()
        self.session = session
    }

    // MARK: - NFCNDEFReaderSessionDelegate

    func readerSession(_ session: NFCNDEFReaderSession, didDetect tags: [NFCNDEFTag]) {
        guard let tag = tags.first else { return }
        session.connect(to: tag) { [weak self] error in
            guard let self else { return }
            if let error {
                self.finish(session, .failure(error))
                return
            }
            // Monta um NDEF message com um único record de texto
            // carregando o transport_field. O leitor (pagador)
            // reconhece o prefixo PIXOFFv1 e roteia para o fluxo mcpix.
            let textPayload = NFCNDEFPayload.wellKnownTypeTextPayload(
                string: self.payload,
                locale: Locale(identifier: "en")
            )
            guard let textPayload else {
                self.finish(session, .failure(NFCError.encodingFailed))
                return
            }
            let message = NFCNDEFMessage(records: [textPayload])
            tag.writeNDEF(message) { writeError in
                if let writeError {
                    self.finish(session, .failure(writeError))
                } else {
                    self.finish(session, .success(()))
                }
            }
        }
    }

    func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        // Sessão encerrada (timeout, cancelamento, ou após escrita).
        // Se ainda não reportamos resultado, propaga o erro.
        if onResult != nil {
            finish(session, .failure(error))
        }
    }

    // Required pela delegate, mas no fluxo de escrita não usamos a
    // leitura de mensagens — só descoberta de tags acima.
    func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {}

    private func finish(_ session: NFCNDEFReaderSession, _ result: Result<Void, Error>) {
        let cb = onResult
        onResult = nil
        if case .success = result {
            session.alertMessage = "Enviado ao terminal."
        }
        session.invalidate()
        cb?(result)
    }
}

enum NFCError: Error {
    case unavailable
    case encodingFailed
}
